File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Dogs/LoadNextDogsPageUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class LoadNextDogsPageUseCase {
    public static let pageSize = 200

    private let dogsRepository: DogsRepository

    public func callAsFunction() -> AnyPublisher<DogsPage, DogsError> {
        dogsRepository.dogsFeed
            .first()
            .setFailureType(to: DogsError.self)
            .flatMap { [dogsRepository] feed in
                dogsRepository.getDogs(offset: feed.dogs.count, limit: Self.pageSize)
            }
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/dogs/LoadNextDogsPageUseCase.kt
package com.perrystreet.woof.usecase.dogs

import com.perrystreet.woof.models.dog.DogsPage
import com.perrystreet.woof.repositories.dogs.DogsRepository
import io.reactivex.rxjava3.core.Single
import org.koin.core.annotation.Factory

@Factory
class LoadNextDogsPageUseCase(
    private val dogsRepository: DogsRepository,
) {
    operator fun invoke(): Single<DogsPage> =
        dogsRepository.dogsFeed
            .firstOrError()
            .flatMap { feed -> dogsRepository.getDogs(offset = feed.dogs.size, limit = PageSize) }

    companion object {
        const val PageSize = 200
    }
}
